Set up dev server and webpack

ckousik opened the pull request:
```
new file:   Dockerfile
new file:   Dockerfile-0
modified:   README.md
modified:   package.json
new file:   src/client/index.jsx
new file:   src/dev-server.js
new file:   webpack.config.js
```

Commit:
Set up dev server and webpack
	new file:   Dockerfile
	new file:   Dockerfile-0
	modified:   README.md
	modified:   package.json
	new file:   src/client/index.jsx
	new file:   src/dev-server.js
	new file:   webpack.config.js

jonasfj commented on Dockerfile-0:
```diff
@@ -0,0 +1,3 @@
+FROM ubuntu
+RUN apt-get update
+RUN apt-get install -y g++ gdb make
```
Why two docker files... we can combine them into one... docker should cache the steps anyways... just make sure to order them correctly.

jonasfj commented on src/dev-server.js:
```diff
@@ -0,0 +1,89 @@
+  await kill();
+  console.log("Killed");
+  return new Promise(function(resolve, reject) {
+    return resolve();
```
Async functions always returns a promise... no need for this... it does nothing...

jonasfj replied:
If you have a value and you don't know if it is a promise or not... you can do:
`Promise.resolve(run()).then(...`

jonasfj commented on src/dev-server.js:
```diff
@@ -0,0 +1,89 @@
+    throw new Error("Docker is not running");
+  }
+  //Set up the docker server
+  docker = new Docker({socketPath:DOCKER_SOCKET});
```
space, please :)

jonasfj replied:
Also don't you need `let docker = ...`

jonasfj commented on src/dev-server.js:
```diff
@@ -0,0 +1,89 @@
+  }
+  //Set up the docker server
+  docker = new Docker({socketPath:DOCKER_SOCKET});
+  await docker.pull('ubuntu');
```
No need... you have `gdp-test` locally...

What we could do is... we could push `gdb-test` to `ckousik/gdb-test` if you create an account on `hub.docker.com`...
Then if we pull from there using `await docker.pull('ckousik/gdb-test');` we don't have to do `docker build -t gdb-test` anymore... Only when we want to update the image, it should make the project easier to hack and remove two steps from the readme :)

Note: thne you would have to set: `Image: "ckousik/gdb-test"` longer down... in `createContainer`.

jonasfj commented on src/dev-server.js:
```diff
@@ -0,0 +1,89 @@
+  await docker.pull('ubuntu');
+  container = await docker.createContainer({
+    Image:'gdb-test',
+    cmd: ['sleep','600']
```
Why not just sleep `Infinity` I think sleep actually can do that...

jonasfj commented on src/dev-server.js:
```diff
@@ -0,0 +1,89 @@
+  var app = express();
+  app.set('views',__dirname+'client');
+  app.set('view engine','jsx');
+  app.engine('jsx',reactViews.createEngine());
```
Let's not render jsx on the server... Can we just put a simple dump HTML file into the `client/` folder..

Something like:

``` html
<html>
<script src="....something that webpack built">
<body>...
</html>
```

ckousik commented:
Sorry I wrote this in a hurry. I'll make the changes ASAP.

jonasfj commented on src/dev-server.js:
```diff
@@ -0,0 +1,89 @@
+
+  return new Promise(function(resolve, reject) {
+    dockerServer.on('exit',()=>{
+      return resolve();
```
Trick 1: You can do: `new Promise((resolve, reject) => {..` (use arrow functions...
Trick 2: You can do: `dockerServer.on('exit', reject);` and `dockerServer.on('error', reject);`
Trick 3: You can do: `dockerServer.once('exit', reject);` and `dockerServer.once('error', reject);`

Number 3 doesn't actually do much :) because calling reject multiple times does nothing to change the promise.

jonasfj commented:
This is definitely down the right path...

Let's not render jsx on the server... write a simple HTML file that loads whatever webpack built...
Then let's see if we can't figure out how to run webpack and the test-server at the same time...

Maybe there is a simple way, or maybe we just run two commands in two terminals and that's fine too...

Commit:
fixed async/await stuff.
created path to query port.

	new file:   src/client/index.html
	modified:   src/client/index.jsx
	modified:   src/dev-server.js

reznord commented:
> Let's not render jsx on the server... write a simple HTML file that loads whatever webpack built...
> Then let's see if we can't figure out how to run webpack and the test-server at the same time...

@jonasfj I think we can use quik for this purpose, since there is an option in quik for creating a static html file : https://github.com/satya164/quik#generating-a-sharable-single-html-file 

I hope this helps a lot.
